articles/110_interface_definition.md

```diff
@@ -0,0 +1,214 @@
+---
+layout: default
+title: Interface definition
+permalink: articles/interface_definition.html
+abstract:
+  This article specifies the format to specify the data structures ROS 2 components exchange to interact with each other.
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+
+## Scope
+
+This article specifies the file format to describe data structures which are being used to exchange information between components.
+The data structures are defined in a programming language agnostic way.
+Please see other articles for the mappings to programming language specific types and API.
+
+
+## Overview
+
+A data structure is defined by a set of *fields*.
+The order of the fields is irrelevant.
+Each field is described by a *type* and a *name*.
+
+
+### Messages
+
+A single data structure is called *message*.
+Each message has a *name*.
+Together with the name of the *package* a message can be uniquely identified.
+
+
+### Services
+
+For request / reply style communication the two exchanged data structures are related.
+These pairs of data structures are called *services*
+A service is identified by its *name* and the *package* it is in.
+Each service describes two messages, one for the request data structure, one for the reply data structure.
+
+
+### Field types
+
+The type of a field can be either a primitive type or another data structure.
+Each of these can optionally be a dynamically or statically sized array.
+
+
+#### Primitive field types
+
+The following primitive types are defined:
+
+* `bool`
+* `byte`
+* `char`
+* `float32`, `float64`
+* `int8`, `uint8`
+* `int16`, `uint16`
+* `int32`, `uint32`
+* `int64`, `uint64`
+* `string`
+
+<div class="alert alert-warning" markdown="1">
+  <b>TODO:</b> consider <code>char</code>, <code>wstring</code>, <code>u16string</code>, <code>u32string</code>
+</div>
+
+
+#### Non-primitive field types
+
+Beside the primitive types other messages can be referenced to describe the type of a "complex" field.
+A complex field type is identified by a package and a message name.
+
+
+#### Arrays with static size
+
+A static array has exactly `N` elements of the specified type.
+`N` must be greater than `0`.
+
+
+#### Arrays with dynamic size
+
+A dynamic array can have between `0` and `N` elements of the specified type.
+`N` might not have an upper bound and may only be limited by the memory or other system specific limitations.
+
+
+#### Upper boundaries
+
+The size of *strings* as well as *dynamic arrays* can be limited with an *upper boundary*.
+This enables to preallocate the memory of data structures which use dynamically sized data.
```

Comment:
OK, to be really indirect it could be "this enables the preallocation of memory for data structures which use dynamically sized data..."

I'll stop nitpicking on this point now that I've come up with something I believe to be both gramatically and conceptually correct :)